src/main/java/com/pi4j/devices/hcsr04/HCSR04.java

```diff
@@ -0,0 +1,167 @@
+        this.trigPin = pi4jContext.create(trigConfig);
+
+        var echoConfig = DigitalInput.newConfigBuilder(pi4jContext)
+            .id("echoSignal")
+            .name("Echo signal for HC-SR04")
+            .address(echoPinNumber)
+            .pull(PullResistance.PULL_DOWN);
+        this.echoPin = pi4jContext.create(echoConfig);
+    }
+
+    /**
+     * Alternative constructor that accepts pre-configured DigitalOutput and DigitalInput pins.
+     *
+     * @param trigPin the DigitalOutput instance for the trigger signal
+     * @param echoPin the DigitalInput instance for the echo signal
+     */
+    public HCSR04(DigitalOutput trigPin, DigitalInput echoPin) {
+        this.trigPin = trigPin;
+        this.echoPin = echoPin;
+    }
+
+    /**
+     * Measures the distance to an object by sending an ultrasonic pulse and measuring the
+     * time taken for the pulse to return. The result is returned in centimeters.
+     *
+     * <p>This method sends a 10-microsecond pulse on the trigger pin and then measures
+     * the duration of the echo pulse on the echo pin. The distance is calculated based
+     * on the duration of the echo pulse.
+     *
+     * @return the calculated distance to the object in centimeters, or -1 if the
+     *         measurement is invalid (out of range).
+     */
+    public double measureDistance() {
+        // Send TRIG signal for 10 microseconds
+        trigPin.high();
+        long trigStartTime = System.nanoTime();
+        while (System.nanoTime() - trigStartTime < TRIG_SIGNAL_DURATION_NANOS) {
+            // Busy-wait for 10 microseconds
+        }
+        trigPin.low();
+
+        // Measure echo pulse duration
+        long echoStartTime = waitForEchoSignal(true);
+        long echoEndTime = waitForEchoSignal(false);
+
+        // Calculate pulse duration
+        long pulseDuration = echoEndTime - echoStartTime;
+
+        // Check if pulse duration is within the expected range
+        if (pulseDuration >= MIN_ECHO_DURATION.toNanos() && pulseDuration <= MAX_ECHO_DURATION.toNanos()) {
+            return calculateDistance(echoStartTime, echoEndTime);
+        } else {
+            return -1; // Return -1 for out-of-range or invalid readings
+        }
+    }
+
+    /**
+     * Waits for the echo signal to either go high or low, depending on the specified parameter.
+     *
+     * <p>This method implements a busy-wait with a timeout to prevent infinite waiting.
+     *
+     * @param high {@code true} to wait for the echo signal to go high, {@code false} to wait for it to go low
+     * @return the time in nanoseconds when the echo signal changed, or the current time if timed out
+     */
+    private long waitForEchoSignal(boolean high) {
+        long timeoutStart = System.nanoTime();
+        while ((echoPin.isHigh() != high) && Duration.ofNanos(System.nanoTime() - timeoutStart).compareTo(MAX_ECHO_DURATION) < 0) {
+            // Busy-wait until signal changes or timeout
```

Comment:
// I considered using an event listener approach to handle the HIGH and LOW signals.
However, since these signals could theoretically occur in any order, we would need to carefully synchronize their detection, likely requiring mechanisms like CountDownLatch to ensure correct timing. Ultimately, this would add unnecessary complexity to the code.
Instead, a simpler polling approach with timeouts provides reliable signal detection without introducing added synchronization challenges.

Reply:
thanks, agree with your decision  with your explanation on possible event order